src/main/java/com/mdabrow9/ticketbookingapp/services/ReservationServiceImpl.java

```diff
@@ -0,0 +1,182 @@
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@Service
+@AllArgsConstructor
+public class ReservationServiceImpl implements ReservationService
+{
+    private final ReservationRepository reservationRepository;
+    private final SeatReservationRepository seatReservationRepository;
+    private final ScreeningRepository screeningRepository;
+    private final SeatRepository seatRepository;
+    private final TicketTypeRepository ticketTypeRepository;
+
+
+    @Transactional
+    @Override
+    public ReservationResponseDTO createReservation(Long[] seatId, Long screeningId, String name, String surname, ReservationTicket[] reservationTickets)
+    {
+        Reservation reservation = new Reservation();
+        //screening
+        Screening screening = getScreening(screeningId);
+        reservation.setScreening(screening);
+
+        if(screening.getStartTime().isBefore(LocalDateTime.now().plusMinutes(15)))
+        {
+            throw new ConflictException("Seats can be booked at latest 15 minutes before the screening begins.");
+        }
+
+
+        //seats
+        List<Seat> seats = getSeats(seatId, screening);
+
+        //checking single seat slots
+        checkSingleSeatSlots(seats, screening);
+        //saving
+        reservation.setSeatReservations(seats.stream()
+                .map(e -> {
+                    SeatReservation s = new SeatReservation();
+                    s.setSeat(e);
+                    s.setScreening(screening);
+                    s.setReservation(reservation);
+                    return s;
+                })
+                .collect(Collectors.toList()));
+
+        //name and surname
+        Pattern compiledSurnamePattern = Pattern.compile("(([AĄBCĆDEĘFGHIJKLŁMNŃOÓPRSŚTUWVXYZŹŻ])[aąbcćdeęfghijklłmnńoóprsśtuwvxyzźż]+)(-([AĄBCĆDEĘFGHIJKLŁMNŃOÓPRSŚTUWVXYZŹŻ])[aąbcćdeęfghijklłmnńoóprsśtuwvxyzźż]+)?");
+        Pattern compiledNamePattern = Pattern.compile("(([AĄBCĆDEĘFGHIJKLŁMNŃOÓPRSŚTUWVXYZŹŻ])[aąbcćdeęfghijklłmnńoóprsśtuwvxyzźż]+)");
+
+
+        if(!compiledNamePattern.matcher(name).matches() || name.length() < 3)
+        {
+            throw new ConflictException("Incorrect client name");
+        }
+        if(!compiledSurnamePattern.matcher(surname).matches() || surname.length() < 3)
+        {
+            throw new ConflictException("Incorrect client surname");
+        }
+        reservation.setCustomerName(name);
+        reservation.setCustomerSurname(surname);
+
+        //tickets
+
+        List<TicketType> tickets = new ArrayList<>();
+
+        for(ReservationTicket item: reservationTickets)
+        {
+            Optional<TicketType> ticketTypeOptional = ticketTypeRepository.findByName(item.getName());
+            if(ticketTypeOptional.isEmpty())
```

Comment:
dlaczego wołanie isEmpty albo isPresent a potem get jest uważane za antypattern? jak to zrobić lepiej z optionalami?